### PR TITLE
revised category backend

### LINKED ISSUE #355: [CategoryBundle] Fix the skipped tests
Currently many of the tests in category bundle are skipped.

I have done this because the CategoryBundle has become broken by default and needs to be repaired, meanwhile we need the tests to be green.

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | yes
| Deprecations? | yes
| Fixed tickets | fixes #355
| Related issues/PRs | -
| License | MIT
| Documentation PR | related to https://github.com/sulu/sulu-docs/pull/263

#### What's in this PR?

This PR merges the changes from #2895 and #2848.
Additionally, the CategoryManager is refactored to return doctrine-entites instead of api-entities.

Summary:

_**refactored category backend**_ (#2895)
- Category api now enables patch requests (preperation for moving categories, #2691)
- Category api respects datagrid expand parameter to expand paths to children
- The category tree in the frontend is lazy loaded now to avoid loading big trees on initialize
- Cleaned up CategoryRepository
- Cleaned up CategoryManager
- Fully tested CategoryController
- Using bundle specific Exceptions in CategoryManager now
- Fixed category assignement for accounts

_**implemented category-bundle entities extensible**_ (#2848)
The following entities were implemented extensible by using the PersistenceBundle:
- Category
- CategoryMeta
- CategoryTranslation
- Keyword

While adapting dependent files, the strategy for setting translations and metas to category-api objects was changed, because entities should not create instance of other entities. 

_**additional changes**_
The CategoryManager was refactored, to return doctine-entities instead of api-entities to avoid BC breaks.

#### BC Breaks

_**API:**_
`/categories`: renamed parameter `parent` which accepted an id to `rootKey` which accepts a key
`/categories/{key}/children` was replaced with `/categories/{id}/children`

_**Classes:**_
`Category\CategoryRepositoryInterface` moved to `Entity\CategoryRepositoryInterface` 
`Category\KeywordRepositoryInterface` moved to `Entity\KeywordRepositoryInterface`
`Category\Exception\KeywordIsMultipleReferencedException` moved to `Exception\KeywordIsMultipleReferencedException` 
`Category\Exception\KeywordNotUniqueException` moved to `Exception\KeywordNotUniqueException` 
`Category\Exception\KeyNotUniqueException` was replaced with `Exception\CategoryKeyNotUniqueException` 

_**Methods:**_
Removed: `Api\Category::setName($name)`
Replacement: `Api\Category::setTranslation(CategoryTranslationInterface $translation)`
Reason: The api-entity cannot create a translation-entity as the translation-entity is implemented extensible.

Removed: `Api\Category::setMeta($metaArrays)`
Replacement: `Api\Category::setMeta($metaEntities)`
Reason: The api-entity cannot create a meta-entity as the meta-entity is implemented extensible.

#### Deprecations

_**Methods:**_
Deprecated: `CategoryRepositoryInterface::findCategoryByIds(array $ids)`
Replacement: `CategoryRepositoryInterface::findCategoriesByIds(array $ids)`

Deprecated: `CategoryRepositoryInterface::findCategories($parent = null, $depth = null, $sortBy = null, $sortOrder = null)`
Replacement: `CategoryRepositoryInterface::findChildrenCategoriesByParentId($parentId = null)`

Deprecated: `CategoryRepositoryInterface::findChildren($key, $sortBy = null, $sortOrder = null)`
Replacement: `CategoryRepositoryInterface::findChildrenCategoriesByParentKey($parentKey = null)`

Deprecated: `CategoryManagerInterface::find($parent = null, $depth = null, $sortBy = null, $sortOrder = null)`
Replacement: `CategoryManagerInterface::findChildrenByParentId($parentId = null)`

Deprecated: `CategoryManagerInterface::findChildren($key, $sortBy = null, $sortOrder = null)`
Replacement: `CategoryManagerInterface::findChildrenByParentKey($parentKey = null)`

_**Container Parameters/Definitions:**_
Deprecated: `sulu_category.entity.category` 
Replacement: `sulu.model.category.class` 

Deprecated: `sulu_category.entity.keyword`
Replacement: `sulu.model.keyword.class`

Deprecated: `sulu_category.category_repository`
Replacement: `sulu.repository.category`

Deprecated: `sulu_category.keyword_repository`
Replacement: `sulu.repository.keyword`

#### Extensibility
Every association of the `Category` entity must be of the type `CategoryInterface` to ensure extensibility
Every association of the `CategoryTranslation` entity must be of the type `CategoryTranslationInterface` to ensure extensibility
Every association of the `CategoryMeta` entity must be of the type `CategoryMetaInterface` to ensure extensibility
Every association of the `Keyword` entity must be of the type `KeywordInterface` to ensure extensibility

### CODE REVIEW

**Comment by nnatter** [comment]
@wachterjohannes

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryMetaInterface.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Entity;
+
+interface CategoryMetaInterface
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryMetaInterface.php]: phpdoc

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryMetaRepository.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Entity;
+
+use Sulu\Component\Persistence\Repository\ORM\EntityRepository;
+
+class CategoryMetaRepository extends EntityRepository implements CategoryMetaRepositoryInterface
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryMetaRepository.php]: phpdoc

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryTranslationInterface.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Entity;
+
+use Doctrine\Common\Collections\Collection;
+use Sulu\Component\Persistence\Model\AuditableInterface;
+use Sulu\Component\Security\Authentication\UserInterface;
+
+interface CategoryTranslationInterface extends AuditableInterface
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryTranslationInterface.php]: phpdoc

File: src/Sulu/Bundle/CategoryBundle/Entity/CategoryTranslationRepository.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\CategoryBundle\Entity;
+
+use Sulu\Component\Persistence\Repository\ORM\EntityRepository;
+
+class CategoryTranslationRepository extends EntityRepository implements CategoryTranslationRepositoryInterface
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/Entity/CategoryTranslationRepository.php]: phpdoc

File: src/Sulu/Bundle/CategoryBundle/SuluCategoryBundle.php [hunk 5]
```diff
@@ -11,8 +11,24 @@
 
 namespace Sulu\Bundle\CategoryBundle;
 
+use Sulu\Bundle\PersistenceBundle\PersistenceBundleTrait;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\HttpKernel\Bundle\Bundle;
 
 class SuluCategoryBundle extends Bundle
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/CategoryBundle/SuluCategoryBundle.php]: phpdoc

**Pull request merged**